src/odemis/gui/comp/viewport.py

```diff
@@ -854,84 +860,124 @@ def setView(self, view, tab_data):
+        if proj:
+            logging.debug("Connecting %s to ARRawViewport", proj)
+            # get polarization/polarimetry positions and set callback function
+            # viewport associated to the projection should have only polarization or polarimetry VA
+            if hasattr(proj, "polarization"):
+                logging.debug("Connecting %s to polarization VA", proj)
+                name = "Polarization"
+                choices = self._stream.polarization.choices
+                if choices == frozenset(POL_POSITIONS):
+                    choices = POL_POSITIONS  # use the ordered tuple to keep legend ordered later
+                cur_value = self._stream.polarization.value
+                self.update_legend_choices(choices, cur_value, name)
+                self._projection.polarization.subscribe(self.on_va_change)
+                self.bottom_legend.Show(True)
+            elif hasattr(proj, "polarimetry"):
+                logging.debug("Connecting %s to polarimetry VA", proj)
+                name = "Polarimetry"
+                cur_value = self._stream.polarimetry.value
+                choices = collections.OrderedDict()
+                choices["Stokes parameters detector plane"] = (MD_POL_DS0, MD_POL_DS1, MD_POL_DS2, MD_POL_DS3)
+                choices["Stokes parameters sample plane"] = (MD_POL_S0, MD_POL_S1, MD_POL_S2, MD_POL_S3)
+                choices["Electrical field amplitudes"] = (MD_POL_EPHI, MD_POL_ETHETA, MD_POL_EX, MD_POL_EY, MD_POL_EZ)
+                choices["Degrees of polarization"] = (MD_POL_DOP, MD_POL_DOLP, MD_POL_DOCP, MD_POL_UP)
+
+                self.update_legend_choices(choices, cur_value, name)
+                self._projection.polarimetry.subscribe(self.on_va_change)
+                self.bottom_legend.Show(True)
+            # if no polarization/polarimetry VA or if no stream (= None)
+            else:
+                # clear legend and also automatically drop callback
+                self.bottom_legend.clear()
+                self.bottom_legend.Show(False)
+            self.Layout()
         else:
-            # clear legend and also automatically drop callback
-            self.bottom_legend.clear()
-            self.bottom_legend.Show(False)
-        self.Layout()
+            logging.info("No stream to show found")
+            self.clear()  # Remove legend ticks and clear plot
 
-    def on_legend_pol_change(self, pol):
+    def on_legend_change(self, pos):
         """
-        Called when the user changes the polarization in the legend.
-        Changes the polarization VA.
+        Called when the user changes the value of the VA displayed in the legend.
+        Changes the VA connected to legend.
+        :param pos: (str) The position to be set on the VA.
         """
-        self.stream.polarization.value = pol
+        if hasattr(self._projection, "polarization"):
+            self._stream.polarization.value = pos
+        if hasattr(self._projection, "polarimetry"):
```

Comment:
The two `if` cannot happen in the same call, right? Then write `elif` in the second one, to make this clearer.